runtime/compiler/runtime/StatisticsThread.cpp

```diff
@@ -0,0 +1,175 @@
+ *******************************************************************************/
+
+#include "runtime/StatisticsThread.hpp"
+#include "env/VMJ9.h"
+#include "runtime/CompileService.hpp"
+#include "control/JITaaSCompilationThread.hpp"
+
+TR_StatisticsThread::TR_StatisticsThread()
+   : _statisticsThread(NULL), _statisticsThreadMonitor(NULL), _statisticsOSThread(NULL),
+   _statisticsThreadAttachAttempted(false), _statisticsThreadExitFlag(false), _statisticsFrequency(TR::Options::getStatisticsFrequency())
+   {
+   }
+
+TR_StatisticsThread * TR_StatisticsThread::allocate()
+   {
+   TR_StatisticsThread * statisticsThread = new (PERSISTENT_NEW) TR_StatisticsThread();
+   return statisticsThread;
+   }
+
+static int32_t J9THREAD_PROC statisticsThreadProc(void * entryarg)
+   {
+   J9JITConfig * jitConfig = (J9JITConfig *) entryarg;
+   J9JavaVM * vm = jitConfig->javaVM;
+
+   TR_StatisticsThread *statisticsThread = ((TR_JitPrivateConfig*)(jitConfig->privateConfig))->statisticsThread;
+   UDATA samplingPeriod    = std::max(static_cast<UDATA>(TR::Options::_minSamplingPeriod), jitConfig->samplingFrequency);
+
+   J9VMThread *statThread = NULL;
+   PORT_ACCESS_FROM_JAVAVM(vm);
+
+   int rc = vm->internalVMFunctions->internalAttachCurrentThread(vm, &statThread, NULL,
+                                  J9_PRIVATE_FLAGS_DAEMON_THREAD | J9_PRIVATE_FLAGS_NO_OBJECT |
+                                  J9_PRIVATE_FLAGS_SYSTEM_THREAD | J9_PRIVATE_FLAGS_ATTACHED_THREAD,
+                                  statisticsThread->getStatisticsOSThread());
+   statisticsThread->getStatisticsThreadMonitor()->enter();
+   statisticsThread->setAttachAttempted(true);
+   if (rc == JNI_OK)
+      statisticsThread->setStatisticsThread(statThread);
+
+   statisticsThread->getStatisticsThreadMonitor()->notifyAll();
+   statisticsThread->getStatisticsThreadMonitor()->exit();
+
+   if (rc != JNI_OK)
+      {
+      return JNI_ERR; // attaching the JITaaS Statistics thread failed
+      }
+
+   j9thread_set_name(j9thread_self(), "JITaaS Server Statistics Thread");
+
+   TR::CompilationInfo *compInfo = TR::CompilationInfo::get(jitConfig);
+   TR::PersistentInfo *persistentInfo = compInfo->getPersistentInfo();
+   uint64_t crtTime = j9time_current_time_millis();
+   uint64_t lastStatistics = crtTime;
+
+   persistentInfo->setStartTime(crtTime);
+   persistentInfo->setElapsedTime(0);
+   while(!statisticsThread->getStatisticsThreadExitFlag())
+      {
+      while(!statisticsThread->getStatisticsThreadExitFlag() && j9thread_sleep_interruptable((IDATA) samplingPeriod, 0) == 0)
+         {
+         crtTime = j9time_current_time_millis();
+         persistentInfo->setElapsedTime(crtTime - persistentInfo->getStartTime());
+
+         if ((statisticsThread->getStatisticsFrequency()) && ((crtTime - lastStatistics) > statisticsThread->getStatisticsFrequency()))
```

Comment:
It occurred to me that the default value for `getStatisticsFrequency` is 0, so stats may be printed every single time. Am I missing something? We agreed that 0 means no printing at all.

Same for samplingFrequency. If that is 0 there should be no statistics thread created.